### PR TITLE
SC-254 Marketplace Product Registration 

### PR DESCRIPTION
**Pull request by brucehoff**
This PR extends the Synapse login app' to support AWS Marketplace registration.
It introduces a new URI, /subscribe, that Marketplace will invoke (HTTP Post) when a user buys our product.
The app' will redirect to Synapse to login, as normal.
When authentication is complete the app' will have (1) the user's Synapse identity and (2) the Marketplace subscription token (which is exchanged for a customer ID).
The app' will bind the Synapse user ID and Marketplace customer ID together, writing them into a DynamoDB record.
The app' will allow each Syn user id to be bound to a single Marketplace customer ID.
After creating the association, the user is redirected to Service Catalog, as usual.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -18,6 +18,9 @@ SESSION_TIMEOUT_SECONDS=43200
 SESSION_NAME_CLAIMS=userid
 SESSION_TAG_CLAIMS=sub,userid,user_name
 REDIRECT_URIS=https://.....,https://....
+MARKETPLACE_PRODUCT_CODE=xxxxxx
```
Comment (zaro0508) [inline README.md]: you'll need to create a PR in Sage-Bionetworks/synapse-login-aws-infra repo to add these params and please make this PR a dependency of that one.

Comment (zaro0508) [inline README.md]: i'm wondering if this naming scheme would work if we were to offer multiple products in the marketplace?  maybe it would be better to set a more specific?  name, like `MARKETPLACE_PROD_CODE_SC`?

File: pom.xml [hunk 2]
```diff
@@ -129,31 +115,8 @@
 				<version>2.3</version>
 				<configuration>
 					<archiveClasses>true</archiveClasses>
-					<webResources>
-						<!-- in order to interpolate version from pom into appengine-web.xml -->
```
Comment (zaro0508) [inline pom.xml:133]: why is this being removed?  i think the /about endpoint depends on this?

File: src/main/webapp/WEB-INF/web.xml [hunk 3]
```diff
@@ -13,6 +13,7 @@
 		<url-pattern>/synapse</url-pattern>
 		<url-pattern>/</url-pattern>
 		<url-pattern>/health</url-pattern>
+		<url-pattern>/subscribe</url-pattern>
```
Comment (zaro0508) [inline src/main/webapp/WEB-INF/web.xml:17]: hmm, where's the `/about` endpoint?  maybe we missed adding it here when we added it?

File: src/main/java/synapseawsconsolelogin/Auth.java [hunk 4]
```diff
@@ -162,17 +207,47 @@ public String getAuthorizeUrl() {
...
+	
+	/*
+	 * Initiate the OAuth login flow for the case in which a user is subscribing to Service Catalog
+	 * as a Marketplace product.  Passes the temporary 'marketplace token' as the OAuth state parameter.
+	 */
+	public void handleSubscribe(HttpServletRequest req, HttpServletResponse resp) throws IOException {
+		String awsMarketPlaceToken = req.getHeader("x-amzn-marketplace-token");
+		if (StringUtils.isEmpty(awsMarketPlaceToken)) {
+			String errorMessage = "Missing x-amzn-marketplace-token header";
+			logger.log(Level.WARNING, errorMessage);
```
Comment (zaro0508) [inline src/main/java/synapseawsconsolelogin/Auth.java]: shouldn't this be `Level.ERROR`?

File: src/main/java/synapseawsconsolelogin/Auth.java [hunk 5]
```diff
@@ -323,6 +400,39 @@ private static void handleException(Exception e, HttpServletResponse resp) throw
...
 		}
 	}
+	
+	/*
+	 * return true if subscription is successful or none needed; false if Synapse user linked to another AWS customer id
+	 */
+	public boolean registerCustomer(HttpServletRequest req, HttpServletResponse resp, String userId) throws IOException {
+		// get the 'state' request parameter
+		String urlEncodedAwsMarketPlaceToken = req.getParameter(STATE);
+		if (StringUtils.isEmpty(urlEncodedAwsMarketPlaceToken)) {
```
Comment (zaro0508) [inline src/main/java/synapseawsconsolelogin/Auth.java:410]: could you please explain why this would result in either successful subscription or non needed?

File: src/test/java/synapseawsconsolelogin/DynamoDbHelperTest.java [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+import com.amazonaws.services.dynamodbv2.model.DescribeTableResult;
+import com.amazonaws.services.dynamodbv2.model.KeySchemaElement;
+import com.amazonaws.services.dynamodbv2.model.KeyType;
+import com.amazonaws.services.dynamodbv2.model.ResourceNotFoundException;
+import com.amazonaws.services.dynamodbv2.model.ScalarAttributeType;
+
+/*
+ * Note: We run this test once, in a local build, to verify that the access to the DynamoDB table is correct.
+ * The test is 'ignored' when checked in so that the CI/CD system does not have to create or delete an actual
+ * table in AWS.
```
Comment (zaro0508) [inline src/test/java/synapseawsconsolelogin/DynamoDbHelperTest.java]: could we make this a unit test (with mocks) so we can run it on every change?  i think that's more useful than a run once test.  Also I prefer to not commit tests that don't get run by the CI system.

File: src/main/java/synapseawsconsolelogin/Auth.java [hunk 7]
```diff
@@ -364,7 +474,22 @@ private void doGetIntern(HttpServletRequest req, HttpServletResponse resp)
...
+			
+			String userId = claims.get(USER_ID_CLAIM_NAME, String.class);
+
+			if (!registerCustomer(req, resp, userId) ) {
+				// previously subscribed to the Marketplace product with this Synapse account and a different AWS customer id.
+				resp.setContentType("text/plain");
+				resp.setCharacterEncoding(UTF8);
+				resp.setStatus(400);
+				PrintWriter out = resp.getWriter();
+				out.print("You have already subscribed to this AWS Marketplace product.");
```
Comment (zaro0508) [inline src/main/java/synapseawsconsolelogin/Auth.java]: should we provide a message telling the user what they should do now?  maybe provide link back to synapse login so they can login and start using the product?

File: src/main/java/synapseawsconsolelogin/Auth.java [hunk 8]
```diff
@@ -150,8 +189,14 @@ public Auth() {
...
 	}
 
-	public String getAuthorizeUrl() {
-		Set<String> allClaims = new TreeSet<String>(getSessionClaimNames());
+	/**
+	 * @param state a value to be returned to this application upon successful login (as defined by the OAuth 'code flow')
+	 * @return the URL to redirect the client to for OAuth login.
+	 */
+	public String getAuthorizeUrl(String state) throws UnsupportedEncodingException {
+		// userid claim is used by this application so it always must be included in the list of claims requested from Synapse
```
Comment (zaro0508) [inline src/main/java/synapseawsconsolelogin/Auth.java:197]: does this comment mean that we don't need to include it in the `SESSION_NAME_CLAIMS` property anymore?  sounds like an update to the doc is needed for this change?

File: src/main/java/synapseawsconsolelogin/Auth.java [hunk 9]
```diff
@@ -426,8 +551,10 @@ public String executeHttpGet(String urlString) throws IOException {
 			PrintWriter out = resp.getWriter();
 			out.print(o.toString());
 			out.flush();
+		} else if (uri.equals(SUBSCRIBE_URI)) {
+			handleSubscribe(req, resp);
```
Comment (zaro0508) [inline src/main/java/synapseawsconsolelogin/Auth.java:568]: just wondering why this is in here and in `doPost`?  handleSubscribe seems like it would be more of POST instead of a GET request.

File: src/main/java/synapseawsconsolelogin/Auth.java [hunk 10]
```diff
@@ -323,6 +400,39 @@ private static void handleException(Exception e, HttpServletResponse resp) throw
...
+		}
+		// write userId + customer ID to Synapse table
+		// look up userId in Synapse table.  Is it already there with another customerIdentifier?
+
+		String existingCustomerId = dynamoDbHelper.getMarketplaceCustomerIdForUser(userId);
+		if (existingCustomerId==null) {
+			dynamoDbHelper.addMarketplaceId(userId, resolveCustomerResult.getProductCode(), resolveCustomerResult.getCustomerIdentifier());
+		} else {
+			if (existingCustomerId.equals(resolveCustomerResult.getCustomerIdentifier())) {
+				// already registered with the existing customer ID, nothing more to do
```
Comment (zaro0508) [inline src/main/java/synapseawsconsolelogin/Auth.java:432]: could we log this?

File: src/main/java/synapseawsconsolelogin/Auth.java [hunk 11]
```diff
@@ -323,6 +400,39 @@ private static void handleException(Exception e, HttpServletResponse resp) throw
...
+		ResolveCustomerResult resolveCustomerResult = marketplaceMeteringHelper.resolveCustomer(awsMarketPlaceToken);
+		if (StringUtils.isNotEmpty(expectedProductCode) && !expectedProductCode.equals(resolveCustomerResult.getProductCode())) {
+			throw new RuntimeException("Expected product code "+expectedProductCode+" but found "+resolveCustomerResult.getProductCode());
+		}
+		// write userId + customer ID to Synapse table
+		// look up userId in Synapse table.  Is it already there with another customerIdentifier?
+
+		String existingCustomerId = dynamoDbHelper.getMarketplaceCustomerIdForUser(userId);
+		if (existingCustomerId==null) {
+			dynamoDbHelper.addMarketplaceId(userId, resolveCustomerResult.getProductCode(), resolveCustomerResult.getCustomerIdentifier());
```
Comment (zaro0508) [inline src/main/java/synapseawsconsolelogin/Auth.java:429]: Would this work for the use case when the same synapse ID is used to register from multiple AWS accounts?

For example:
register for the SC from AWS account A using synapse ID 12345
register for the SC from AWS account B using synapse ID 12345

**Review by tthyer — APPROVED** [review]

**Review by zaro0508 — CHANGES_REQUESTED** [review]
please make this PR a dependency of https://github.com/Sage-Bionetworks/scipool-infra/pull/214

File: README.md (hunk 1 above)
Comment (brucehoff) [reply README.md]: Done!

Comment (brucehoff) [reply README.md]: Good suggestion: It's not yet clear to me how this approach will generalize to multiple products but doing this might help avoid "painting ourselves into a corner."

File: pom.xml (hunk 2 above)
Comment (brucehoff) [reply pom.xml:133]: The comment suggests it there for use by 'appengine' which we no longer use.  I restored the deleted code (minus the misleading comment).

File: src/main/java/synapseawsconsolelogin/Auth.java (hunk 8 above)
Comment (brucehoff) [reply src/main/java/synapseawsconsolelogin/Auth.java:197]: > does this comment mean that we don't need to include it in the `SESSION_NAME_CLAIMS` property anymore? 

No, removing the claim from `SESSION_NAME_CLAIMS` would stop it from being appended to the assumed role (and break downstream code).

> sounds like an update to the doc is needed for this change?

I added a note to the ReadMe explaining that `userid` will always be requested from Synapse.

File: src/main/java/synapseawsconsolelogin/Auth.java (hunk 5 above)
Comment (brucehoff) [reply src/main/java/synapseawsconsolelogin/Auth.java:410]: Yes:  We pass through here both for regular login and when registering a customer who purchased
SC as a product.  We differentiate the two cases by the presence of the marketplace token.
If absent, it's a simple login and we skip the registration step.

I added this explanation in the comments.

File: src/main/java/synapseawsconsolelogin/Auth.java (hunk 11 above)
Comment (brucehoff) [reply src/main/java/synapseawsconsolelogin/Auth.java:429]: No.  Below you will see that if the same Synapse user has already purchased the product from another AWS account we display the message, "You have already subscribed to this AWS Marketplace product." and stop.

File: src/main/java/synapseawsconsolelogin/Auth.java (hunk 10 above)
Comment (brucehoff) [reply src/main/java/synapseawsconsolelogin/Auth.java:432]: yes!

File: src/main/java/synapseawsconsolelogin/Auth.java (hunk 7 above)
Comment (brucehoff) [reply src/main/java/synapseawsconsolelogin/Auth.java]: Yes, good suggestion.

File: src/main/java/synapseawsconsolelogin/Auth.java (hunk 9 above)
Comment (brucehoff) [reply src/main/java/synapseawsconsolelogin/Auth.java:568]: Excellent question:  Our redirector returns the status 302 Found.  I am not certain how the HTTP client used by AWS will respond to a 302. It might issue a POST to the redirected URL or it might issue a GET.  (See https://en.wikipedia.org/wiki/HTTP_302). This bit of code makes our system robust in the face of unexpected behavior by AWS.

File: src/main/webapp/WEB-INF/web.xml (hunk 3 above)
Comment (brucehoff) [reply src/main/webapp/WEB-INF/web.xml:17]: added it

File: src/test/java/synapseawsconsolelogin/DynamoDbHelperTest.java (hunk 6 above)
Comment (brucehoff) [reply src/test/java/synapseawsconsolelogin/DynamoDbHelperTest.java]: Done

File: README.md (hunk 1 above)
Comment (zaro0508) [inline README.md]: please update to `MARKETPLACE_PRODUCT_CODE_SC` to match change in the code.

File: README.md [hunk 12]
```diff
@@ -48,13 +51,22 @@ https://rest-docs.synapse.org/rest/org/sagebionetworks/repo/model/oauth/OIDCClai
 used to define tags in the AWS session.  The tags are names `synapse-`<claim name>, where <claim name> is the name of the claim given in the config file.
 
 The `SESSION_NAME_CLAIMS` config is also a comma separated list of claims, but used to define the session name, as a colon delimited list of claim values. For example: setting `SESSION_NAME_CLAIMS=userid,email` will display
-`ServiceCatalogEndusers/1234567:joe.smith@gmail.com` in AWS. 
+`ServiceCatalogEndusers/1234567:joe.smith@gmail.com` in AWS.
+
+Note:  The list of claims requested from Synapse is the union of the two lists, `SESSION_TAG_CLAIMS` and `SESSION_NAME_CLAIMS`, plus the `userid` claim, which this application uses itself.
```
Comment (zaro0508) [inline README.md:56]: > plus the `userid` claim, which this application uses itself.

Does this mean that `userid` is required to be on the   SESSION_NAME_CLAIMS list?  if so could we reword it to be more explicit?

**Review by zaro0508 — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (brucehoff) [reply README.md]: done

File: README.md (hunk 12 above)
Comment (brucehoff) [reply README.md:56]: > Does this mean that userid is required to be on the SESSION_NAME_CLAIMS list?

No. The app' takes the union of SESSION_NAME_CLAIMS, SESSION_TAG_CLAIMS and `userid` to get the list of things to ask Synapse for.  The configuration of SESSION_NAME_CLAIMS and SESSION_TAG_CLAIMS need not include userid.

**Review by zaro0508 — APPROVED** [review]
